heartbeat/monitors/active/tcp/tcp.go

```diff
@@ -39,113 +43,219 @@ func init() {
+	return createWithResolver(cfg, monitors.StdResolver)
+}
+
+// Custom resolver is useful for tests against hostnames locally where we don't want to depend on any
+// hostnames existing in test environments
+func createWithResolver(
+	cfg *common.Config,
+	resolver monitors.Resolver,
+) (jobs []jobs.Job, endpoints int, err error) {
+	jc, err := makeJobFactory(cfg, resolver)
+	if err != nil {
 		return nil, 0, err
 	}
 
-	tls, err := tlscommon.LoadTLSConfig(config.TLS)
+	jobs, err = jc.makeJobs()
 	if err != nil {
 		return nil, 0, err
 	}
 
-	defaultScheme := "tcp"
-	if tls != nil {
-		defaultScheme = "ssl"
+	return jobs, len(jc.endpoints), nil
+}
+
+// jobFactory is where most of the logic here lives. It provides a common context around
+// the complex logic of executing a TCP check.
+type jobFactory struct {
+	config        Config
+	tlsConfig     *tlscommon.TLSConfig
+	defaultScheme string
+	endpoints     []endpoint
+	dataCheck     dataCheck
+	resolver      monitors.Resolver
+}
+
+func makeJobFactory(commonCfg *common.Config, resolver monitors.Resolver) (*jobFactory, error) {
+	jf := &jobFactory{config: DefaultConfig, resolver: resolver}
+	err := jf.loadConfig(commonCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return jf, nil
+}
+
+// loadConfig parses the YAML config and populates the jobFactory fields.
+func (jf *jobFactory) loadConfig(commonCfg *common.Config) (err error) {
+	if err = commonCfg.Unpack(&jf.config); err != nil {
+		return err
 	}
 
-	schemeHosts, err := collectHosts(&config, defaultScheme)
+	jf.tlsConfig, err = tlscommon.LoadTLSConfig(jf.config.TLS)
 	if err != nil {
-		return nil, 0, err
+		return err
 	}
 
-	timeout := config.Timeout
-	validator := makeValidateConn(&config)
+	jf.defaultScheme = "tcp"
+	if jf.tlsConfig != nil {
+		jf.defaultScheme = "ssl"
+	}
 
-	for scheme, eps := range schemeHosts {
-		schemeTLS := tls
-		if scheme == "tcp" || scheme == "plain" {
-			schemeTLS = nil
-		}
+	jf.endpoints, err = makeEndpoints(jf.config.Hosts, jf.config.Ports, jf.defaultScheme)
+	if err != nil {
+		return err
+	}
 
-		db, err := dialchain.NewBuilder(dialchain.BuilderSettings{
-			Timeout: timeout,
-			Socks5:  config.Socks5,
-			TLS:     schemeTLS,
-		})
-		if err != nil {
-			return nil, 0, err
-		}
+	jf.dataCheck = makeDataCheck(&jf.config)
 
-		epJobs, err := dialchain.MakeDialerJobs(db, scheme, eps, config.Mode,
-			func(event *beat.Event, dialer transport.Dialer, addr string) error {
-				return pingHost(event, dialer, addr, timeout, validator)
-			})
-		if err != nil {
-			return nil, 0, err
+	return nil
+}
+
+// makeJobs returns the actual schedulable jobs for this monitor.
+func (jf *jobFactory) makeJobs() ([]jobs.Job, error) {
+	var jobs []jobs.Job
+	for _, endpoint := range jf.endpoints {
+		for _, url := range endpoint.perPortURLs() {
+			endpointJob, err := jf.makeEndpointJob(url)
+			if err != nil {
+				return nil, err
+			}
+			jobs = append(jobs, wrappers.WithURLField(url, endpointJob))
```

Comment:
I wonder if this is a change in semantics introduced here or a while ago already. By creating a job per complete URL, we will have each Job do it's own DNS request. Originally only one job was created to do the initial DNS lookup, with follow up tasks per IP:Port combination. On Windows/MacOS DNS is cached automatically, but we have had incidents on Linux overloading DNS servers, due to missing caching.

Reply:
I took a look at the code here, and I do think we can fix this so we restore the behavior before this patch. However, I agree with you that it's not a clear win for `ports` because that code is tricky. I think for `mode: all` it still makes sense to keep the behavior, because more people use that (and it provides benefits for icmp/tcp/http not just tcp).

However, for TCP, given that this is a very rarely used feature AFAICT I don't think the code complexity is worth maintaining the single DNS lookup. Additionally, there are workarounds in the rare case where DNS load is too high. One could always install a caching resolver on the same box as heartbeat in the worst case.